src/core/mac/sub_mac.cpp

```diff
@@ -591,10 +611,30 @@ void SubMac::HandleTransmitDone(TxFrame &aFrame, RxFrame *aAckFrame, Error aErro
     {
         mTransmitRetries++;
         aFrame.SetIsARetransmission(true);
+
+#if OPENTHREAD_CONFIG_MAC_ADD_DELAY_ON_NO_ACK_ERROR_BEFORE_RETRY
+        if (aError == kErrorNoAck)
+        {
+            SetState(kStateDelayBeforeRetx);
+            StartTimerForBackoff(mRetxDelayBackOffExponent);
```

Comment:
The timer is started in HandleTransmitDone instead of BeginTransmit right? Is that ok? In case of getting a Data Poll is there a possibility an indirect frame can be sent in-between direct frame retransmissions. Would that case be handled in this approach?

Reply:
> The timer is started in HandleTransmitDone instead of BeginTransmit right? Is that ok? 

Yes. Correct. We want to add delay before we start the next retx which will have its own CSMA process (which may be done by radio platform if it supports it or if not will be done by `SubMac` code in `StartCsmaBackoff()`.  In the existing code, when we determine that frame still needs to be retx, we would immediately call `StartCsmaBackoff()` from here (and in the case radio does the CSMA, `StartCsmaBackoff()` would directly call `BeginTransmit()`).

> In case of getting a Data Poll is there a possibility an indirect frame can be sent in-between direct frame retransmissions.

That is not possible (with the implementation in general). `SubMac` layer deals with one frame tx request at a time and then it is up to the next layer (`Mac` and other layer) to deal with data poll and direct vs indirect messages. 
`Mac` itself has the concept of `Operation`. At any point a single `Operation` is active and being executed. It keeps track of other requests and remembers the pending operations. Once the current operation is finished, it determines the highest priority next operation to start with.